### PR TITLE
Podfile Refactor

### LINKED ISSUE #840: Clean up Podfile syntax
#### Background

_Note: This text will be updated to reflect the most recent proposal_

Currently CocoaPods:

- Skips target definitions without dependencies as the user might not want the implicit root target definition (which automatically links with the first target) and there is no other way to prevent its integration.
- Marks a target as exclusive if it has a different platform from the parent not inheriting any dependency.
- Doesn't have a way to specify what should be inherited from the parent target definition (nothing, only the headers, the dependencies). This is evident in the tests target which currently need to be marked as exclusive, however this set up only works because CocoaPods makes the headers of all targets visible (exactly for this reason).
- Doesn't allow to specify that a target should not be integrated in the Podfile. It is only possible to specify that whole Podfile should not be integrated (i.e. no workspace) with a command line option.

#### Proposal

- [ ] Add a new `abstract_target!` attribute to the target definition.
  - Skip the installation of the implicit root target definition if it doesn't have any dependency (temporary see below).
  - Otherwise never skip the installation of a target unless marked abstract.
- [ ] Add a new `do_not_integrate!` root attribute which would replace the command line option (`--no-integrate` of pod install and pod update).
- [ ] Inherit the compatible dependencies from the parent target definition if the platforms do not match.
- [ ] Add a new `inherit` attribute to the target definition which would support 3 values (true, false, :headers).
  - Phase out the `exclusive` option of the target definition.
- [ ] Drop any form of integration of the implicit target definition by CP 1.0. Adding a warning immediately.
- [ ] Drop the options parameter used for new target definitions (the has used after the name and before the block) in favor of statements inside the block.
  - This would accelerate the phasing out of the [exclusive](http://docs.cocoapods.org/podfile.html#target) option.
- [ ] Deprecate the `link_with` attribute. Every integrated target should be identified by the name of the target definition common dependencies can be declared in dedicated targets. This would not only simplify the DSL but also the implementation of the CP analyzer and user project integrator.

#### Points which require Feedback

- The name of the `abstract_target!` and of the `do_not_integrate!`
- Wether inheritance by nesting targets definitions should be dropped for an attribute (`inherits_from`, se example below).

###### Headers only inheritance

Note that By CocoaPods 1.0 the `abstract_target!` call can be dropped from the root as it will become the default behaviour.

```ruby
abstract_target!
pod 'CocoaAsyncSocket', '0.0.1'
pod 'CocoaLumberjack', '>=1.6'

target 'App_1'
  # deps
end

target 'App_2'
  # deps
end
```

---

###### Abstract targets

```ruby
target 'MacApp'
  pod 'AFNetworking'
 
  target 'MacAppTests' do
     use_headers_of_parent
     # dependencies
  end
end
```

--

###### Alternative headers only inheritance

In this scenario nesting would be disabled.

```ruby
target 'MacApp'
  pod 'AFNetworking'
end

target 'MacAppTests' do
  use_headers_of 'MacApp' 
  # use_headers_of 'MacApp_1', 'MacApp_2'
  # dependencies
end
```

---

###### Multi platform inheritance

```ruby
abstract_target!
pod 'CocoaAsyncSocket', '0.0.1'
pod 'CocoaLumberjack', '>=1.6'

target 'iOSApp'
  platform :ios
  # deps
end

target 'MacApp'
  platform :osx 
  # deps
end
```

### PR DESCRIPTION
**Pull request by segiddins**
This is a work in progress.
Depends upon https://github.com/CocoaPods/Core/pull/281.
Closes #840.
Closes https://github.com/CocoaPods/CocoaPods/issues/4673.
Closes https://github.com/CocoaPods/CocoaPods/issues/3617.
Closes https://github.com/CocoaPods/CocoaPods/issues/1249.
Closes https://github.com/CocoaPods/CocoaPods/issues/1626.
Closes https://github.com/CocoaPods/CocoaPods/issues/4001.

- [x] Documentation
- [x] Sane commits
- [x] CHANGELOGs
- [x] More tests
  - [x] Updated integration specs for new DSL
  - [x] Covering search path inheritance actually being used
  - [x] Abstract-only pod validation
  - [x] Otherwise-empty search-paths inheritance target can compile
  - [x] Empty podfile works 
- [x] Pull install options from Podfile, remove from CLI flags
- [x] Make empty target definitions integrate as expected

### CODE REVIEW

File: lib/cocoapods/installer/analyzer/target_inspector.rb [hunk 1]
```diff
@@ -102,18 +102,9 @@ def compute_project_path
         #
         def compute_targets(user_project)
```
Comment (segiddins) [inline lib/cocoapods/installer/analyzer/target_inspector.rb]: make this not return an array

**Comment by orta** [comment]
Cool, I've started looking at this [in Eigen.](https://github.com/artsy/eigen/tree/orta-cp_podfile) 

After making some ARAnalytics changes, I get this exception message:

NoMethodError - undefined method `deployment_target=' for #<Xcodeproj::Project::Object::PBXNativeTarget:0x007fd8fddc5e28>
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:119:in `block (2 levels) in add_resources_bundle_targets'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:115:in `each'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:115:in `block in add_resources_bundle_targets'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:114:in `each'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:114:in `add_resources_bundle_targets'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:20:in `block in install!'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/user_interface.rb:144:in `message'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer/target_installer/pod_target_installer.rb:17:in `install!'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:634:in `block (2 levels) in install_libraries'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:631:in `each'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:631:in `block in install_libraries'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/user_interface.rb:144:in `message'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:630:in `install_libraries'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:166:in `block in generate_pods_project'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/user_interface.rb:63:in `section'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:163:in `generate_pods_project'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/installer.rb:115:in `install!'
/Users/orta/.gem/ruby/2.1.3/gems/cocoapods-keys-1.6.0/lib/plugin.rb:99:in `install!'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/command/project.rb:67:in `run_install_with_update'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/command/project.rb:149:in `run'
/Users/orta/.gem/ruby/2.1.3/gems/claide-0.9.1/lib/claide/command.rb:312:in `run'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/lib/cocoapods/command.rb:48:in `run'
/Users/orta/.gem/ruby/2.1.3/bundler/gems/CocoaPods-cd49b659b4bc/bin/pod:44:in `<top (required)>'
/Users/orta/.gem/ruby/2.1.3/bin/pod:23:in `load'
/Users/orta/.gem/ruby/2.1.3/bin/pod:23:in `<main>'
```

A bunch of times I got this error too:, I saw it with ARAnalytics earlier. both of which are custom `:git` pods.

```
[!] The `ARGenericTableViewController` pod failed to validate due to 1 error:
    - ERROR | name: The name of the spec should match the name of the file.
```

To give you full details:
```
Fetching external sources
-> Pre-downloading: `ARGenericTableViewController` from `https://github.com/orta/ARGenericTableViewController.git`, commit `f4c9028754558e196ab9c5111dcd3a4229c5bedf`
 > Git download
 > Git download
     $ /usr/bin/git clone https://github.com/orta/ARGenericTableViewController.git /var/folders/d9/p68r8wr93h55rxl_k281js2h0000gp/T/d20151213-57845-11zve7y
     Cloning into '/var/folders/d9/p68r8wr93h55rxl_k281js2h0000gp/T/d20151213-57845-11zve7y'...
   $ /usr/bin/git checkout --quiet f4c9028754558e196ab9c5111dcd3a4229c5bedf
  > Copying ARGenericTableViewController from `/Users/orta/Library/Caches/CocoaPods/Pods/External/ARGenericTableViewController/3d7c241d9d2076172d293292b511ef72` to `Pods/ARGenericTableViewController`
-> Pre-downloading: `JSDecoupledAppDelegate` from `https://github.com/orta/JSDecoupledAppDelegate.git`, commit `cd72cdfd4543a26b01dc229d6265737331696435`
  > Copying JSDecoupledAppDelegate from `/Users/orta/Library/Caches/CocoaPods/Pods/External/JSDecoupledAppDelegate/3fff64b9451b61ff8e0424ccc722b355` to `Pods/JSDecoupledAppDelegate`
[!] The `JSDecoupledAppDelegate` pod failed to validate due to 1 error:
    - ERROR | name: The name of the spec should match the name of the file.
```

I checked the folders, they didn't have a Podspec - but neither did any of the folders in my cache dir.

**Comment by segiddins** [comment]
Are you using Xcodeproj master?

**Comment by orta** [comment]
Not then, am now, I'm not getting past the matching spec name at all, if I comment out the above example, I get another pod with it instead.

```
   eigen   cp_podfile ✘ ✹  bundle exec pod install
Updating local specs repositories
Analyzing dependencies
Pre-downloading: `JSDecoupledAppDelegate` from `https://github.com/orta/JSDecoupledAppDelegate.git`, commit `cd72cdfd4543a26b01dc229d6265737331696435`
[!] The `JSDecoupledAppDelegate` pod failed to validate due to 1 error:
    - ERROR | name: The name of the spec should match the name of the file.
```

**Comment by segiddins** [comment]
Ok cool will look into that one tonight.

**Comment by segiddins** [comment]
@orta can't reproduce the issue using this branch and eigen master

**Comment by orta** [comment]
Updated everything on [this eigen branch](https://github.com/artsy/eigen/tree/orta-cp_podfile) - 

```
RuntimeError - [Xcodeproj] Consistency issue: build setting `IPHONEOS_DEPLOYMENT_TARGET` has multiple values: `{"Release"=>"4.3", "Debug"=>"4.3", "Beta"=>"8.0", "Store"=>"8.0", "Demo"=>"8.0"}`
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/Xcodeproj-0dbacb0eaf3e/lib/xcodeproj/project/object/native_target.rb:79:in `common_resolved_build_setting'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/Xcodeproj-0dbacb0eaf3e/lib/xcodeproj/project/object/native_target.rb:129:in `deployment_target'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/Xcodeproj-0dbacb0eaf3e/lib/xcodeproj/project/object/native_target.rb:172:in `add_build_configuration'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:156:in `block (3 levels) in add_resources_bundle_targets'
```

**Comment by orta** [comment]
Same in [Energy](https://github.com/artsy/energy/tree/orta-cp_podfile)

```
RuntimeError - [Xcodeproj] Consistency issue: build setting `IPHONEOS_DEPLOYMENT_TARGET` has multiple values: `{"Release"=>"6.0", "Debug"=>"6.0", "Beta"=>"8.0"}`
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/Xcodeproj-0dbacb0eaf3e/lib/xcodeproj/project/object/native_target.rb:79:in `common_resolved_build_setting'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/Xcodeproj-0dbacb0eaf3e/lib/xcodeproj/project/object/native_target.rb:129:in `deployment_target'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/Xcodeproj-0dbacb0eaf3e/lib/xcodeproj/project/object/native_target.rb:172:in `add_build_configuration'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:156:in `block (3 levels) in add_resources_bundle_targets'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:155:in `each'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:155:in `block (2 levels) in add_resources_bundle_targets'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:129:in `each'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:129:in `block in add_resources_bundle_targets'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:128:in `each'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:128:in `add_resources_bundle_targets'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/installer/target_installer/pod_target_installer.rb:20:in `block in install!'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-2a6d86a82b43/lib/cocoapods/user_interface.rb:144:in `message'
```

**Comment by segiddins** [comment]
@orta it'll be fixed once I rebase against master

**Comment by orta** [comment]
no worries, just wanting you to feel like it's being tested ;)

File: lib/cocoapods/installer/installation_options.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        end
+      end
+
+      option :clean, true
+      option :deduplicate_targets, true
+      option :deterministic_uuids, true
+      option :integrate_targets, true
+      option :lock_pod_sources, true
+
+      module Mixin
```
Comment (segiddins) [inline lib/cocoapods/installer/installation_options.rb]: make the delegation opt-in

**Comment by segiddins** [comment]
Empty target definitions now work!

File: lib/cocoapods/installer/installation_options.rb (hunk 2 above)
Comment (segiddins) [inline lib/cocoapods/installer/installation_options.rb]: this is done

**Comment by orta** [comment]
Not getting a complete Pod install yet on [eigen](https://github.com/artsy/eigen/tree/orta-cp_podfile):

```
TypeError - no implicit conversion of String into Integer
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-010cda1ba964/lib/cocoapods/installer/installation_options.rb:7:in `[]'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-010cda1ba964/lib/cocoapods/installer/installation_options.rb:7:in `from_podfile'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-010cda1ba964/lib/cocoapods/installer/installation_options.rb:53:in `block in delegate_installation_options'
/Users/orta/.rvm/gems/ruby-2.1.3/bundler/gems/CocoaPods-010cda1ba964/lib/cocoapods/installer.rb:229:in `block in create_analyzer'
```

**Comment by segiddins** [comment]
@orta I can't reproduce that issue on the branch you pointed me to, but I think it should now be fixed anyways.

**Comment by orta** [comment]
Interesting, well, I re-ran it and it worked :tada:

File: lib/cocoapods/installer/installation_options.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        def Mixin.included(mod)
+          mod.send(:attr_accessor, :installation_options)
+
+          def mod.delegate_installation_options(&blk)
```
Comment (segiddins) [inline lib/cocoapods/installer/installation_options.rb]: do these by extending an inner module, so we can document them

File: lib/cocoapods/downloader.rb [hunk 4]
```diff
@@ -27,7 +27,8 @@ module Downloader
     def self.download(
       request,
       target,
-      cache_path: !Config.instance.skip_download_cache && Config.instance.clean? && Config.instance.cache_root + 'Pods'
+      can_cache: true,
+      cache_path: can_cache && !Config.instance.skip_download_cache && Config.instance.cache_root + 'Pods'
     )
       if cache_path
```
Comment (segiddins) [inline lib/cocoapods/downloader.rb]: `&& can_cache`

**Comment by neonichu** [comment]
LGTM :+1:

Also did some testing with a few projects and seems to work well. Awesome work, @segiddins :tada:

**Comment by segiddins** [comment]
:ship:

**Pull request merged**